[TG-9017] Extend java-assume-inputs-interval to a list of intervals

romainbrenguier opened the pull request:
This allow using a list of intervals for the option rather than just a single one.
The interval_uniont class is moved from the test-gen repository to this one since it can be used in both.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Add an interval_union class

This was already used in the test-gen repository but will also be useful
for JBMC assume-inputs-interval option.

romainbrenguier committed:
Move split_string documentation to header

Makes it easier to find.

romainbrenguier committed:
Use interval_uniont for assume-inputs-interval

Instead of a single interval, this makes it more flexible in its use.

romainbrenguier committed:
Test with list for --java-assume-inputs-interval

This tests that the option works for a list of interval rather than just
a single one.

romainbrenguier committed:
Add unit test for interval union

This is moved from the test-gen repository.

peterschrammel commented:
💡 it would also be good if equal lower and upper bounds are encoded as equality so that constant propagation picks it up.

romainbrenguier committed:
Add an as_singleton method to interval_uniont

This to do specific things in the special case of a singleton.

romainbrenguier committed:
Replace assumption by assignment for singletons

When the union of intervals is a singleton we use an assignement instead
of a aassumption, this will allow constant propagation.

romainbrenguier committed:
Add a test for singleton in assume-inputs-interval

This is to check that constant propagation takes place.

romainbrenguier committed:
Add an all_integers method for interval_uniont

Makes it easy to construct the set of all integers.

romainbrenguier committed:
Relax inputs interval tests for error messages

The message is now slightly different as the format has been extended.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e58faadefd4cb7329b4011a90e1cfc7831f6dd70).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129815154

smowton commented on src/util/interval_union.cpp:
```diff
@@ -0,0 +1,236 @@
+      ++it2;
+    }
+  }
+  while(it1 != intervals.end() || it2 != other.intervals.end())
```
Invariant here that it1 and it2 are both either `end()` or have `lower_set`?

romainbrenguier replied:
I guess you mean `upper_set`?

smowton replied:
I actually meant `lower_set`, because we just executed a prelude specifically for intervals open at the lower end -- I expected at this point any interval is closed at the bottom

romainbrenguier replied:
:+1: I got it now

smowton commented on src/util/interval_union.cpp:
```diff
@@ -0,0 +1,236 @@
+      !back.lower_set ||
+        ((it1 == intervals.end() || it1->lower >= back.lower) &&
+         (it2 == other.intervals.end() || it2->lower >= back.lower)),
+      "intervals are processed in order");
```
Phrase as a "should-be" statement, an specify ordering by `lower`

smowton commented on src/util/interval_union.cpp:
```diff
@@ -0,0 +1,236 @@
+  if(other.intervals.empty())
+    return *this;
+
+  interval_uniont result{};
```
Give a prose sketch of the algorithm

smowton commented on src/util/interval_union.cpp:
```diff
@@ -0,0 +1,236 @@
+  return output.str();
+}
+
+exprt interval_uniont::contains_expr(const exprt &e) const
```
`get_contains_expr`? `contains_expr` suggests it would return a boolean indicating whether `to_constant_expr(e)` fell within the specified interval

romainbrenguier replied:
`get_contains_expr` makes me think of a getter and not something that is constructing an expression. `contains_expr` follows the model of `equal_expr` which doesn't return a Boolean but an expression, but I agree this can be ambiguous, however the return type makes it clear.

smowton replied:
`make_contains_expr`?

smowton commented on src/util/interval_union.h:
```diff
@@ -0,0 +1,86 @@
+  optionalt<mp_integer> as_singleton() const;
+
+private:
+  /// Interval stored in order so that each interval is strictly below the
```
Be explicit -- that implies non-overlapping, right? So there can be at most one open interval at the bottom and one at the top?

smowton commented on src/util/interval_union.h:
```diff
@@ -0,0 +1,86 @@
+private:
+  /// Interval stored in order so that each interval is strictly below the
+  /// following one.
+  std::vector<intervalt> intervals;
```
Suggest having a `validate` method that checks this conforms to expectations, and INVARIANT'ing that after the `from_string` method for example

smowton reviewed (approved):
LGTM, some suggestions

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7e3984f0de962da77c03f5b157989ef73e19da41).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129863008
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 85cf470727961d391425d3281a49b1d322145431).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/130683974
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

segun3d reviewed (commented):
Checked developer regression tests and also did some exploratory testing
QA Approved

svorenova commented on src/util/interval_union.h:
```diff
@@ -0,0 +1,85 @@
+  /// greater or equal to \p value.
+  static interval_uniont smaller_or_equal(const mp_integer &value);
+
+  NODISCARD interval_uniont
```
:pick: Missing documentation

majakusber replied:
:pick: Missing documentation

majakusber commented on src/util/interval_union.h:
```diff
@@ -0,0 +1,85 @@
+  NODISCARD interval_uniont
+  make_intersection(const interval_uniont &other) const;
+
+  NODISCARD interval_uniont make_union(const interval_uniont &other) const;
```
:pick: Missing documentation

majakusber commented on src/util/Makefile:
```diff
@@ -28,6 +28,7 @@ SRC = allocate_objects.cpp \
       get_module.cpp \
       identifier.cpp \
       ieee_float.cpp \
+      interval_union.cpp \
```
From the commit message it sounds like this will be removed from test-gen, in a bump presumably?

romainbrenguier replied:
https://github.com/diffblue/test-gen/pull/3970

majakusber commented on src/util/string_utils.cpp:
```diff
@@ -37,16 +37,6 @@ std::string strip_string(const std::string &s)
   return s.substr(i, (j-i+1));
 }
 
-/// Given a string s, split into a sequence of substrings when separated by
```
:pick: The commit message says `strip_string` but the function is called `split_string`

majakusber commented on unit/util/interval_union.cpp:
```diff
@@ -0,0 +1,55 @@
+  REQUIRE(first.make_intersection(second).to_string() == "[:-15][5:5][8:10]");
+
+  REQUIRE(first.make_intersection(third).is_empty());
+  REQUIRE(first.make_intersection(third).to_string() == "[]");
```
:idea: What would happen if you do `from_string("[1:3][2:4]")`? Does it create `[1:4]`?

romainbrenguier replied:
yes

majakusber commented on jbmc/src/java_bytecode/java_object_factory_parameters.cpp:
```diff
@@ -22,30 +22,21 @@ void java_object_factory_parameterst::set(const optionst &options)
     {
       throw invalid_command_line_argument_exceptiont(
-        "interval must be of the form [int:int], [int:] or [:int]",
+        "argument must be a sequence of intervals of the form [int:int], [int:]"
```
`must be a comma-separated sequence..`

majakusber commented on jbmc/src/java_bytecode/java_object_factory_parameters.cpp:
```diff
@@ -22,30 +22,21 @@ void java_object_factory_parameterst::set(const optionst &options)
+    if(interval->is_empty())
+    {
+      throw invalid_command_line_argument_exceptiont(
+        "interval is empty, lower limit cannot be bigger than upper limit",
```
:grey_question: Is this the only situation when the interval ends up empty? I cannot think of any other but you may have a deeper insight into it.

romainbrenguier replied:
I guess if you give it an empty string, but I'm not sure the option parser would accept that...

majakusber commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1076,10 +1076,19 @@ void java_object_factoryt::gen_nondet_init(
+      if(auto singleton = interval.as_singleton())
+      {
+        assignments.add(
+          code_assignt{expr, from_integer(*singleton, expr.type())});
```
:grey_question: Can this actually override the original value? Does that matter?

romainbrenguier replied:
the original value is nondet, so it doesn't matter

majakusber commented on jbmc/regression/jbmc/assume-inputs-interval/arg_assume_singleton.desc:
```diff
@@ -0,0 +1,22 @@
+^warning: ignoring
+--
+Check when --java-assume-inputs-interval specifies a singleton, constant
+propagation can take place.
```
Please explain that constant propagation is checked by checking for 0 VCCs

majakusber commented on jbmc/regression/jbmc/assume-inputs-interval/arg_assume_invalid2.desc:
```diff
@@ -4,7 +4,7 @@ My.class
 ^SIGNAL=0$
 ^Invalid User Input
-^Reason: interval must be of the form
+^Reason: argument must be .* of the form
```
:pick: Can this be anything else than ` argument must be a sequence of intervals of the form`?

majakusber reviewed (commented):
Few minor requests/questions, otherwise looks great :+1:
